File: src/install/resolve.ts
import os from "node:os";
import path from "node:path";
import { execFileSync } from "node:child_process";

import { InstallError } from "./errors.js";
import { uniqueValues } from "../utils/collections.js";

export const AGENTS = [
  "codex",
  "claude",
  "portable",
  "vscode",
  "copilot",
  "amp",
  "goose",
  "opencode",
  "factory",
  "cursor",
] as const;

export const SCOPES = ["repo", "user", "admin", "cwd", "parent"] as const;

export type Agent = (typeof AGENTS)[number];
export type Scope = (typeof SCOPES)[number];

type ScopeResolvers = Partial<Record<Scope, (cwd: string) => string>>;

export function resolveRepoRoot(cwd: string): string {
  try {
    const out = execFileSync("git", ["rev-parse", "--show-toplevel"], {
      cwd,
      encoding: "utf8",
    }).trim();
    if (out) return out;
  } catch {
    // fall through
  }
  return cwd;
}

function configRoot(): string {
  return process.env.XDG_CONFIG_HOME ?? path.join(os.homedir(), ".config");
}

const scopeResolversByAgent: Record<Agent, ScopeResolvers> = {
  codex: {
    repo: (cwd) => path.join(resolveRepoRoot(cwd), ".codex/skills"),
    cwd: (cwd) => path.join(cwd, ".codex/skills"),
    parent: (cwd) => path.join(path.resolve(cwd, ".."), ".codex/skills"),
    user: () => {
      const root = process.env.CODEX_HOME ?? path.join(os.homedir(), ".codex");
      return path.join(root, "skills");
    },
    admin: () => "/etc/codex/skills",
  },
  claude: {
    repo: (cwd) => path.join(resolveRepoRoot(cwd), ".claude/skills"),
    user: () => path.join(os.homedir(), ".claude/skills"),
  },
  portable: {
    repo: (cwd) => path.join(resolveRepoRoot(cwd), ".agents/skills"),
    user: () => path.join(configRoot(), "agents/skills"),
  },
  vscode: {
    repo: (cwd) => path.join(resolveRepoRoot(cwd), ".github/skills"),
  },
  copilot: {
    repo: (cwd) => path.join(resolveRepoRoot(cwd), ".github/skills"),
  },
  amp: {
    repo: (cwd) => path.join(resolveRepoRoot(cwd), ".agents/skills"),
    user: () => path.join(configRoot(), "agents/skills"),
  },
  goose: {
    repo: (cwd) => path.join(resolveRepoRoot(cwd), ".agents/skills"),
    user: () => path.join(configRoot(), "agents/skills"),
  },
  opencode: {
    repo: (cwd) => path.join(resolveRepoRoot(cwd), ".opencode/skill"),
    user: () => path.join(configRoot(), "opencode/skill"),
  },
  factory: {
    repo: (cwd) => path.join(resolveRepoRoot(cwd), ".factory/skills"),
    user: () => path.join(os.homedir(), ".factory/skills"),
  },
  cursor: {
    repo: (cwd) => path.join(resolveRepoRoot(cwd), ".cursor/skills"),
  },
};

export function assertAgent(value: string): Agent {
  if ((AGENTS as readonly string[]).includes(value)) {
    return value as Agent;
  }
  throw new InstallError(`Unsupported agent: ${value}`);
}

export function assertScope(value: string): Scope {
  if ((SCOPES as readonly string[]).includes(value)) {
    return value as Scope;
  }
  throw new InstallError(`Unsupported scope: ${value}`);
}

export function supportedScopesForAgent(agent: Agent): Scope[] {
  return Object.keys(scopeResolversByAgent[agent]) as Scope[];
}

export function sharedScopesForAgents(agents: Agent[]): Scope[] {
  const uniqueAgents = uniqueValues(agents);
  if (uniqueAgents.length === 0) {
    return [];
  }

  const first = uniqueAgents[0];
  return supportedScopesForAgent(first).filter((scope) =>
    uniqueAgents.every((agent) =>
      supportedScopesForAgent(agent).includes(scope),
    ),
  );
}

export function assertSupportedAgentScopes(
  agents: Agent[],
  scopes: Scope[],
): void {
  for (const agent of uniqueValues(agents)) {
    const supported = supportedScopesForAgent(agent);
    for (const scope of uniqueValues(scopes)) {
      if (!supported.includes(scope)) {
        throw new InstallError(`Unsupported agent/scope: ${agent} ${scope}`);
      }
    }
  }
}

export function resolveSkillsRoot(
  agent: Agent,
  scope: Scope,
  cwd: string,
): string {
  const resolver = scopeResolversByAgent[agent][scope];
  if (!resolver) {
    throw new InstallError(`Unsupported agent/scope: ${agent} ${scope}`);
  }
  return resolver(cwd);
}

File: python/src/skillflag/install/resolve.py
"""Agent/scope destination table (port of typescript/src/install/resolve.ts)."""

from __future__ import annotations

import os
import subprocess
from collections.abc import Callable, Sequence
from pathlib import Path

from skillflag._util import unique_values
from skillflag.install.errors import InstallError

AGENTS = (
    "codex",
    "claude",
    "portable",
    "vscode",
    "copilot",
    "amp",
    "goose",
    "opencode",
    "factory",
    "cursor",
)

SCOPES = ("repo", "user", "cwd")


def resolve_repo_root(cwd: str) -> str:
    try:
        out = subprocess.run(
            ["git", "rev-parse", "--show-toplevel"],
            cwd=cwd,
            capture_output=True,
            text=True,
            check=True,
        ).stdout.strip()
        if out:
            return out
    except (OSError, subprocess.CalledProcessError):
        pass
    return cwd


def _home() -> str:
    return str(Path.home())


def _config_root() -> str:
    value = os.environ.get("XDG_CONFIG_HOME")
    if value is not None:
        return value
    return os.path.join(_home(), ".config")


def _codex_home() -> str:
    value = os.environ.get("CODEX_HOME")
    if value is not None:
        return value
    return os.path.join(_home(), ".codex")


# Insertion order of scopes matters: it drives the interactive scope listing.
_SCOPE_RESOLVERS: dict[str, dict[str, Callable[[str], str]]] = {
    "codex": {
        "repo": lambda cwd: os.path.join(resolve_repo_root(cwd), ".codex", "skills"),
        "cwd": lambda cwd: os.path.join(cwd, ".codex", "skills"),
        "user": lambda cwd: os.path.join(_codex_home(), "skills"),
    },
    "claude": {
        "repo": lambda cwd: os.path.join(resolve_repo_root(cwd), ".claude", "skills"),
        "user": lambda cwd: os.path.join(_home(), ".claude", "skills"),
    },
    "portable": {
        "repo": lambda cwd: os.path.join(resolve_repo_root(cwd), ".agents", "skills"),
        "user": lambda cwd: os.path.join(_config_root(), "agents", "skills"),
    },
    "vscode": {
        "repo": lambda cwd: os.path.join(resolve_repo_root(cwd), ".github", "skills"),
    },
    "copilot": {
        "repo": lambda cwd: os.path.join(resolve_repo_root(cwd), ".github", "skills"),
    },
    "amp": {
        "repo": lambda cwd: os.path.join(resolve_repo_root(cwd), ".agents", "skills"),
        "user": lambda cwd: os.path.join(_config_root(), "agents", "skills"),
    },
    "goose": {
        "repo": lambda cwd: os.path.join(resolve_repo_root(cwd), ".agents", "skills"),
        "user": lambda cwd: os.path.join(_config_root(), "agents", "skills"),
    },
    "opencode": {
        "repo": lambda cwd: os.path.join(resolve_repo_root(cwd), ".opencode", "skill"),
        "user": lambda cwd: os.path.join(_config_root(), "opencode", "skill"),
    },
    "factory": {
        "repo": lambda cwd: os.path.join(resolve_repo_root(cwd), ".factory", "skills"),
        "user": lambda cwd: os.path.join(_home(), ".factory", "skills"),
    },
    "cursor": {
        "repo": lambda cwd: os.path.join(resolve_repo_root(cwd), ".cursor", "skills"),
    },
}


def assert_agent(value: str) -> str:
    if value in AGENTS:
        return value
    raise InstallError(f"Unsupported agent: {value}")


def assert_scope(value: str) -> str:
    if value in SCOPES:
        return value
    raise InstallError(f"Unsupported scope: {value}")


def supported_scopes_for_agent(agent: str) -> list[str]:
    return list(_SCOPE_RESOLVERS[agent].keys())


def shared_scopes_for_agents(agents: Sequence[str]) -> list[str]:
    unique_agents = unique_values(agents)
    if not unique_agents:
        return []
    first = unique_agents[0]
    return [
        scope
        for scope in supported_scopes_for_agent(first)
        if all(scope in supported_scopes_for_agent(agent) for agent in unique_agents)
    ]


def assert_supported_agent_scopes(agents: Sequence[str], scopes: Sequence[str]) -> None:
    for agent in unique_values(agents):
        supported = supported_scopes_for_agent(agent)
        for scope in unique_values(scopes):
            if scope not in supported:
                raise InstallError(f"Unsupported agent/scope: {agent} {scope}")


def resolve_skills_root(agent: str, scope: str, cwd: str) -> str:
    resolver = _SCOPE_RESOLVERS[agent].get(scope)
    if resolver is None:
        raise InstallError(f"Unsupported agent/scope: {agent} {scope}")
    return resolver(cwd)
